save-core/src/commonMain/kotlin/org/cqfn/save/core/files/ConfigDetector.kt

```diff
@@ -0,0 +1,79 @@
+package org.cqfn.save.core.files
+
+import org.cqfn.save.core.config.TestSuiteConfig
+import org.cqfn.save.core.config.isDefaultSaveConfig
+
+import okio.FileSystem
+import okio.Path
+
+/**
+ * A class that is capable of discovering config files hierarchy.
+ */
+class ConfigDetector {
+    /**
+     * Try to create SAVE config file from [file].
+     *
+     * @param file a [Path] from which SAVE config file should be built.
+     * @return [TestSuiteConfig] or null if no suitable config file has been found.
+     */
+    fun configFromFile(file: Path): TestSuiteConfig? = discoverConfigWithParents(file)
+        ?.also { config ->
+            // fill children for parent configs
+            config.parentConfigs(wihSelf = true).toList().reversed()
+                .zipWithNext().forEach { (parent, child) ->
+                    parent.childConfigs.add(child)
+                }
+            // discover all descendant configs of [config]
+            val locationsFlattened = config.location.parent!!.findAllFilesMatching { it.isDefaultSaveConfig() }.flatten()
+            val configs = mutableListOf(config)
+            locationsFlattened
+                .drop(1)  // because [config] will be discovered too
+                .forEachIndexed { index, path ->
+                    configs.add(
+                        TestSuiteConfig(
+                            "todo: read from file",
+                            "todo: read from file",
+                            path,
+                            configs.find { discoveredConfig ->
+                                discoveredConfig.location ==
+                                        locationsFlattened.take(index + 1).reversed().find { it.parent in path.parents() }!!
+                            }!!
+                        ).also {
+                            it.parentConfig!!.childConfigs.add(it)
```

Comment:
parent config can be null here, I guess

Reply:
No, because it's a loop among descendants of current config, so each of them will have a parent